### ISSUE TITLE
U+11443 NEWA SIGN CANDRABINDU is too big in ⟨𑐬𑑂𑐎𑐶𑑃⟩

### ISSUE THREAD

**Issue by dscorbett** [notofonts/newa#9]
### Font
NotoSansNewa-Regular.otf

### Where the font came from, and when
Site: https://github.com/notofonts/newa/releases/tag/NotoSansNewa-v2.006
Date: 2022-10-14

### Font version
Version 2.006

### Issue
In version 2.006, U+11443 NEWA SIGN CANDRABINDU no longer takes its small form after a repha and U+11436 NEWA VOWEL SIGN I.

### Character data
𑐬𑑂𑐎𑐶𑑃
U+1142C NEWA LETTER RA
U+11442 NEWA SIGN VIRAMA
U+1140E NEWA LETTER KA
U+11436 NEWA VOWEL SIGN I
U+11443 NEWA SIGN CANDRABINDU

### Screenshots
![𑐬𑑂𑐎𑐶𑑃](https://user-images.githubusercontent.com/1124347/195866924-4943095e-bf04-422b-9d0c-8121ceb07cd3.png)
Compare version 2.005:
![𑐬𑑂𑐎𑐶𑑃](https://user-images.githubusercontent.com/1124347/195867068-20072c1d-9c43-42e8-93e5-61b83249b291.png)

**Comment by simoncozens** [comment]
I'll fix this - I had to move some lookups around and it had some fallout on other stuff - but I'm really intrigued about how you found this kind of difference (especially so quickly). Do you have a stock of sample texts to compare? I am wondering what we can learn for our own QA processes.

**Comment by dscorbett** [comment]
I have no test corpus; I just compare the old and new font files. In this case, because all the lookups were rearranged, I ran this:
```bash
otf2fea ~/Library/Fonts/NotoSansNewa-Regular.otf >0.fea
otf2fea ~/Downloads/NotoSansNewa-v2.006/NotoSansNewa/unhinted/otf/NotoSansNewa-Regular.otf >1.fea
diff <(sed 's/\(class\|GSUB\|GPOS\|source: \|Substitution\|Position\|Anchor\|Set\)[0-9]\+/\1#/g' 0.fea | sort) \
    <(sed 's/\(class\|GSUB\|GPOS\|source: \|Substitution\|Position\|Anchor\|Set\)[0-9]\+/\1#/g' 1.fea | sort)
```
A smarter OTL-specific rearrangement-aware diff tool would be nice, but this is good enough to find things worthy of further investigation. I saw a lot of new `sub Candrabindu by Candrabindu.sm;` rules, and I knew that this new version was supposed to fix something about rephas and dependent vowels, so I tried a string with a repha, a dependent vowel, and a candrabindu.

For your own QA process, the closest thing to human intuition is fuzzing: you could run a fuzzer that shapes strings focusing on characters that appear in modified OTL rules or that are mentioned in the release notes. You could also add rules to Font Bakery codifying the common patterns in my bug reports, like “glyphs overlap unexpectedly” or “marks interrupt contextual lookups” or “that is not how USE works”.

**Comment by simoncozens** [comment]
We have implemented quite a bit of this! fontbakery has a shaping check which can shape patterns of glyphs ("base virama base mark") and test for collisions. We also now have diffenator2 which can compare shaped texts between old and new fonts, although it requires a wordlist to compare. siesta can generate sequences based on the rules and encoded codepoints in the font. But putting it all together in a smart way which can generate intelligent rules based on expected failures from the feature code is... a challenge. I think diffenator2 and gathering a corpus of attested sequences has the most potential.

Anyway, I reordered the features here and it seems to be doing the right thing.
